### ISSUE TITLE
Zshcomp needs various improvements including test coverage

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2291]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)| @teratorn reported|
|-|-|
|Trac ID|trac#2291|
|Type|defect|
|Created|2006-12-11 10:04:54Z|

The zsh completion functions for twistd and trial used to tell you all the reactor choices you had, but recent refactorings removed that functionality.

<details><summary>Searchable metadata</summary>

```
trac-id__2291 2291
type__defect defect
reporter__teratorn teratorn
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1165831494000000 1165831494000000
changetime__1166862121000000 1166862121000000
version__None None
owner__teratorn teratorn

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Feel free to try to get this done before Twisted 2.5 is released, but it isn't a release blocking issue.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_2"></a>|@teratorn commented|
|-|-|

No it certainly is not, but I figured I would set the milestone so that at least someone would see it before throwing the release switch.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_3"></a>|@teratorn removed owner|
|-|-|

I've included various other improvements to zsh completion in this branch. Whoever reviews this should ask me about things that don't make sense or need clarification. Will be happy to add more comments or better documentation if that strikes the reviewer as beneficial.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @teratorn|
|-|-|

The branch changes untested code without adding test coverage.  See ReviewProcess for a list of other things which will also cause a branch to fail review. ;)

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_5"></a>|@teratorn set owner to @exarkun|
|-|-|

Yeah, this code was written years ago before the current stricter guidelines.

So, I've added several tests in tests/test_zshcomp.py.

Of note is testBuildAll(). It simply builds all the completions for all the twisted commands. If anyone changes anything to break the build process then we will know about it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @teratorn|
|-|-|

* twisted/test/test_zshcomp.py
  * copyright should be 2006
  * stdlib imports should come before twisted imports
  * ZshcompTestCase should have a docstring
  * testBuildAll should have a docstring
  * testBuildAll should use self.mktemp() to come up with a directory name.  Otherwise, it might collide with another test, or if it is run repeatedly it will collide with itself.
  * testAccumulateMetadata and testAccumulateAdditionalOptions have good docstrings, but they should be reformatted:
```
def test_foo(self):
    """
    docstring
    """
```
  * Likewise, the test method names should be ```test_foo``` instead of ```testFoo```.  All new tests should be named like this, and any existing tests modified should be renamed.
  * testVerifyZshNames should have a docstring
  * TestOptions and TestOptions2 should have docstrings
  * ```opt_nocrash``` and ```opt_difficulty``` don't need ```pass``` in their body, their docstrings satisfy the grammatical requirement that a suite have at least one statement.
 * twisted/python/zshcomp.py
  * SubcommandBuilder should have a docstring, as should its write method
  * MktapBuilder and TwistdBuilder should have docstrings as well
  * Thanks for cleaning up the comment cruft
  * Is the ability to place callables into the various zsh_ dictionaries intended to be a public feature?  Framework code uses it now, but it is undocumented and untested.

I think there needs to be at least some test coverage for the actual ZSH code
generation part of twisted/python/zshcomp.py.  This doesn't need to explicitly
test the code generated for twistd and mktap and so forth: it can test simpler
Options classes. Ideally, there'd be a _ton_ of test coverage: code generation
is hairy at best; _shell_ code generation is bleuch.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_8"></a>|@teratorn set owner to @exarkun|
|-|-|

Thanks a lot for your very thorough review!

I believe I have addressed each point in the current branch.

In particular there is a new test case which actually tests the generated code in the only way I can think of: I generates a function and compares the text output against a known good output. This should catch many regressions that would otherwise cause incorrect output.

Regarding callables in zsh_* dicts: They can only be used in zsh_actions because that is the only place I've wanted to use them, and it seems likely that they won't be needed elsewhere. It has been documented accordingly.

Thanks again :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @teratorn|
|-|-|

Okay, looking better.  A few last things:

  * Clean up the trailing whitespace in zshcomp.py, test_zshcomp.py, and trial.py
  * Wrap at least all new or modified lines in .py files to <80 cols
  * For the zsh_actions function value documentation, please document the arguments the function should take (I guess they don't take any) and the type and meaning of the return value.

Once that's done, feel free to merge.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_10"></a>|@teratorn set status to `closed`|
|-|-|

(In [18979]) Merge updateZshcomp-2291

Author: teratorn
Reviewer: exarkun
Fixes [#2291](https://github.com/twisted/twisted/issues/2291)

Various improvements and enhancements to zshcomp; Add test coverage to zshcomp;
remove trailing whitespace from affected files; update completion functions
to match current code

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_11"></a>|@teratorn set status to `reopened`|
|-|-|

Reopening. A newly added test causes problems on buildslaves that don't have needed libraries installed. E.g. Tkinter.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_12"></a>|@teratorn set status to `closed`|
|-|-|

(In [19005]) Merge updateZshcomp-2291 

Author: teratorn
Reviewer: exarkun
Fixes [#2291](https://github.com/twisted/twisted/issues/2291) 

Various improvements to zshcomp including adding test coverage; remove trailing whitespace from affected files; update completion functions to match current code

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_13"></a>|@teratorn set status to `reopened`|
|-|-|

Reopening. There is still a single regression on one of the win32 buildslaves.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_14"></a>|@teratorn set owner to @glyph<br>@teratorn set status to `new`|
|-|-|

The branch should be good now. Tests pass on the buildslave in question (win32-py2.4-select).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph set owner to @teratorn|
|-|-|

OK.  Looks fine to me.  Tests pass on a local Windows machine.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_17"></a>|@teratorn set status to `closed`|
|-|-|

(In [19035]) Merge updateZshcomp-2291 

Author: teratorn
Reviewer: exarkun, glyph
Fixes [#2291](https://github.com/twisted/twisted/issues/2291) 

Various improvements to zshcomp including adding test coverage; remove trailing whitespace from affected files; update completion functions to match current code
